File: searching_sorting/restaurant_customers/restaurant_customers.py
from sys import stdin

def max_customers(times):
    times.sort()
    m = 0
    cur_customers = 0
    max_func = max
    for i in range(len(times)):
        cur_customers += 1 if times[i][1] == "A" else -1
        m = max_func(m, cur_customers)
    return m

if __name__ == '__main__':
    n = int(stdin.readline())
    times = []
    for _ in range(n):
        a_time, d_time = [int(num) for num in stdin.readline().split()]
        times.append((a_time, "A"))
        times.append((d_time, "D"))
    print(max_customers(times))

File: searching_sorting/restaurant_customers/restaurant_customers.cpp
#include <iostream>
#include <utility>
#include <vector>
#include <algorithm>
using namespace std;

int max_customers(vector<pair <int, char> > times) {
    sort(times.begin(), times.end());
    int max_customers = 0;
    int cur_max = 0;
    for (int i = 0; i < times.size(); i++) {
        cur_max += times[i].second == 'A' ? 1 : -1;
        max_customers = max(cur_max, max_customers);
    }
    return max_customers;
}

int main() {
    int n;
    scanf("%d", &n);
    int a_time, d_time;
    vector <pair <int, char> > times;
    for (int i = 0; i < n; i++) {
        scanf("%d %d", &a_time, &d_time);
        times.push_back(make_pair(a_time, 'A'));
        times.push_back(make_pair(d_time, 'D'));
    }
    cout << max_customers(times) << endl;
}

